// devextreme-angular/bundles/devextreme-angular-ui-file-manager.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/platform-browser'), require('@angular/core'), require('devextreme/ui/file_manager'), require('devextreme-angular/core'), require('devextreme-angular/ui/nested')) :
    typeof define === 'function' && define.amd ? define('devextreme-angular/ui/file-manager', ['exports', '@angular/platform-browser', '@angular/core', 'devextreme/ui/file_manager', 'devextreme-angular/core', 'devextreme-angular/ui/nested'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["devextreme-angular"] = global["devextreme-angular"] || {}, global["devextreme-angular"].ui = global["devextreme-angular"].ui || {}, global["devextreme-angular"].ui["file-manager"] = {}), global.ng.platformBrowser, global.ng.core, global.DxFileManager, global["devextreme-angular"].core, global["devextreme-angular"].ui.nested));
})(this, (function (exports, i2, i0, DxFileManager, i1, nested) { 'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var DxFileManager__default = /*#__PURE__*/_interopDefaultLegacy(DxFileManager);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    function __createBinding(o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    }
    function __exportStar(m, exports) {
        for (var p in m)
            if (p !== "default" && !exports.hasOwnProperty(p))
                exports[p] = m[p];
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (Object.hasOwnProperty.call(mod, k))
                    result[k] = mod[k];
        result.default = mod;
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    /**
     * The FileManager is a UI component that allows users to upload, select, and manage files and directories in different file storages.

     */
    var DxFileManagerComponent = /** @class */ (function (_super) {
        __extends(DxFileManagerComponent, _super);
        function DxFileManagerComponent(elementRef, ngZone, templateHost, _watcherHelper, _idh, optionHost, transferState, platformId) {
            var _this = _super.call(this, elementRef, ngZone, templateHost, _watcherHelper, transferState, platformId) || this;
            _this._watcherHelper = _watcherHelper;
            _this._idh = _idh;
            _this._createEventEmitters([
                { subscribe: 'contentReady', emit: 'onContentReady' },
                { subscribe: 'contextMenuItemClick', emit: 'onContextMenuItemClick' },
                { subscribe: 'contextMenuShowing', emit: 'onContextMenuShowing' },
                { subscribe: 'currentDirectoryChanged', emit: 'onCurrentDirectoryChanged' },
                { subscribe: 'directoryCreated', emit: 'onDirectoryCreated' },
                { subscribe: 'directoryCreating', emit: 'onDirectoryCreating' },
                { subscribe: 'disposing', emit: 'onDisposing' },
                { subscribe: 'errorOccurred', emit: 'onErrorOccurred' },
                { subscribe: 'fileUploaded', emit: 'onFileUploaded' },
                { subscribe: 'fileUploading', emit: 'onFileUploading' },
                { subscribe: 'focusedItemChanged', emit: 'onFocusedItemChanged' },
                { subscribe: 'initialized', emit: 'onInitialized' },
                { subscribe: 'itemCopied', emit: 'onItemCopied' },
                { subscribe: 'itemCopying', emit: 'onItemCopying' },
                { subscribe: 'itemDeleted', emit: 'onItemDeleted' },
                { subscribe: 'itemDeleting', emit: 'onItemDeleting' },
                { subscribe: 'itemDownloading', emit: 'onItemDownloading' },
                { subscribe: 'itemMoved', emit: 'onItemMoved' },
                { subscribe: 'itemMoving', emit: 'onItemMoving' },
                { subscribe: 'itemRenamed', emit: 'onItemRenamed' },
                { subscribe: 'itemRenaming', emit: 'onItemRenaming' },
                { subscribe: 'optionChanged', emit: 'onOptionChanged' },
                { subscribe: 'selectedFileOpened', emit: 'onSelectedFileOpened' },
                { subscribe: 'selectionChanged', emit: 'onSelectionChanged' },
                { subscribe: 'toolbarItemClick', emit: 'onToolbarItemClick' },
                { emit: 'accessKeyChange' },
                { emit: 'activeStateEnabledChange' },
                { emit: 'allowedFileExtensionsChange' },
                { emit: 'contextMenuChange' },
                { emit: 'currentPathChange' },
                { emit: 'currentPathKeysChange' },
                { emit: 'customizeDetailColumnsChange' },
                { emit: 'customizeThumbnailChange' },
                { emit: 'disabledChange' },
                { emit: 'elementAttrChange' },
                { emit: 'fileSystemProviderChange' },
                { emit: 'focusedItemKeyChange' },
                { emit: 'focusStateEnabledChange' },
                { emit: 'heightChange' },
                { emit: 'hintChange' },
                { emit: 'hoverStateEnabledChange' },
                { emit: 'itemViewChange' },
                { emit: 'notificationsChange' },
                { emit: 'permissionsChange' },
                { emit: 'rootFolderNameChange' },
                { emit: 'rtlEnabledChange' },
                { emit: 'selectedItemKeysChange' },
                { emit: 'selectionModeChange' },
                { emit: 'tabIndexChange' },
                { emit: 'toolbarChange' },
                { emit: 'uploadChange' },
                { emit: 'visibleChange' },
                { emit: 'widthChange' }
            ]);
            _this._idh.setHost(_this);
            optionHost.setHost(_this);
            return _this;
        }
        Object.defineProperty(DxFileManagerComponent.prototype, "accessKey", {
            /**
             * Specifies the shortcut key that sets focus on the UI component.
            
             */
            get: function () {
                return this._getOption('accessKey');
            },
            set: function (value) {
                this._setOption('accessKey', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "activeStateEnabled", {
            /**
             * Specifies whether the UI component changes its visual state as a result of user interaction.
            
             */
            get: function () {
                return this._getOption('activeStateEnabled');
            },
            set: function (value) {
                this._setOption('activeStateEnabled', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "allowedFileExtensions", {
            /**
             * Specifies the allowed upload file extensions.
            
             */
            get: function () {
                return this._getOption('allowedFileExtensions');
            },
            set: function (value) {
                this._setOption('allowedFileExtensions', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "contextMenu", {
            /**
             * Configures the context menu settings.
            
             */
            get: function () {
                return this._getOption('contextMenu');
            },
            set: function (value) {
                this._setOption('contextMenu', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "currentPath", {
            /**
             * Specifies the path that is used when the FileManager is initialized.
            
             */
            get: function () {
                return this._getOption('currentPath');
            },
            set: function (value) {
                this._setOption('currentPath', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "currentPathKeys", {
            /**
             * Specifies an array of path keys to the current location.
            
             */
            get: function () {
                return this._getOption('currentPathKeys');
            },
            set: function (value) {
                this._setOption('currentPathKeys', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "customizeDetailColumns", {
            /**
             * Customizes columns in details view. Applies only if itemView.mode is &apos;details&apos;.
            
             */
            get: function () {
                return this._getOption('customizeDetailColumns');
            },
            set: function (value) {
                this._setOption('customizeDetailColumns', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "customizeThumbnail", {
            /**
             * Allows you to provide custom icons to be used as thumbnails.
            
             */
            get: function () {
                return this._getOption('customizeThumbnail');
            },
            set: function (value) {
                this._setOption('customizeThumbnail', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "disabled", {
            /**
             * Specifies whether the UI component responds to user interaction.
            
             */
            get: function () {
                return this._getOption('disabled');
            },
            set: function (value) {
                this._setOption('disabled', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "elementAttr", {
            /**
             * Specifies the global attributes to be attached to the UI component&apos;s container element.
            
             */
            get: function () {
                return this._getOption('elementAttr');
            },
            set: function (value) {
                this._setOption('elementAttr', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "fileSystemProvider", {
            /**
             * Specifies the file system provider.
            
             */
            get: function () {
                return this._getOption('fileSystemProvider');
            },
            set: function (value) {
                this._setOption('fileSystemProvider', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "focusedItemKey", {
            /**
             * Specifies a key of the initially or currently focused item.
            
             */
            get: function () {
                return this._getOption('focusedItemKey');
            },
            set: function (value) {
                this._setOption('focusedItemKey', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "focusStateEnabled", {
            /**
             * Specifies whether the UI component can be focused using keyboard navigation.
            
             */
            get: function () {
                return this._getOption('focusStateEnabled');
            },
            set: function (value) {
                this._setOption('focusStateEnabled', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "height", {
            /**
             * Specifies the UI component&apos;s height.
            
             */
            get: function () {
                return this._getOption('height');
            },
            set: function (value) {
                this._setOption('height', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "hint", {
            /**
             * Specifies text for a hint that appears when a user pauses on the UI component.
            
             */
            get: function () {
                return this._getOption('hint');
            },
            set: function (value) {
                this._setOption('hint', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "hoverStateEnabled", {
            /**
             * Specifies whether the UI component changes its state when a user pauses on it.
            
             */
            get: function () {
                return this._getOption('hoverStateEnabled');
            },
            set: function (value) {
                this._setOption('hoverStateEnabled', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "itemView", {
            /**
             * Configures the file and directory view.
            
             */
            get: function () {
                return this._getOption('itemView');
            },
            set: function (value) {
                this._setOption('itemView', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "notifications", {
            /**
             * Configures notification settings.
            
             */
            get: function () {
                return this._getOption('notifications');
            },
            set: function (value) {
                this._setOption('notifications', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "permissions", {
            /**
             * Specifies actions that a user is allowed to perform on files and directories.
            
             */
            get: function () {
                return this._getOption('permissions');
            },
            set: function (value) {
                this._setOption('permissions', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "rootFolderName", {
            /**
             * Specifies the root directory display name.
            
             */
            get: function () {
                return this._getOption('rootFolderName');
            },
            set: function (value) {
                this._setOption('rootFolderName', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "rtlEnabled", {
            /**
             * Switches the UI component to a right-to-left representation.
            
             */
            get: function () {
                return this._getOption('rtlEnabled');
            },
            set: function (value) {
                this._setOption('rtlEnabled', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "selectedItemKeys", {
            /**
             * Contains an array of initially or currently selected files and directories&apos; keys.
            
             */
            get: function () {
                return this._getOption('selectedItemKeys');
            },
            set: function (value) {
                this._setOption('selectedItemKeys', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "selectionMode", {
            /**
             * Specifies whether a user can select a single or multiple files and directories in the item view simultaneously.
            
             */
            get: function () {
                return this._getOption('selectionMode');
            },
            set: function (value) {
                this._setOption('selectionMode', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "tabIndex", {
            /**
             * Specifies the number of the element when the Tab key is used for navigating.
            
             */
            get: function () {
                return this._getOption('tabIndex');
            },
            set: function (value) {
                this._setOption('tabIndex', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "toolbar", {
            /**
             * Configures toolbar settings.
            
             */
            get: function () {
                return this._getOption('toolbar');
            },
            set: function (value) {
                this._setOption('toolbar', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "upload", {
            /**
             * Configures upload settings.
            
             */
            get: function () {
                return this._getOption('upload');
            },
            set: function (value) {
                this._setOption('upload', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "visible", {
            /**
             * Specifies whether the UI component is visible.
            
             */
            get: function () {
                return this._getOption('visible');
            },
            set: function (value) {
                this._setOption('visible', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxFileManagerComponent.prototype, "width", {
            /**
             * Specifies the UI component&apos;s width.
            
             */
            get: function () {
                return this._getOption('width');
            },
            set: function (value) {
                this._setOption('width', value);
            },
            enumerable: false,
            configurable: true
        });
        DxFileManagerComponent.prototype._createInstance = function (element, options) {
            return new DxFileManager__default["default"](element, options);
        };
        DxFileManagerComponent.prototype.ngOnDestroy = function () {
            this._destroyWidget();
        };
        DxFileManagerComponent.prototype.ngOnChanges = function (changes) {
            _super.prototype.ngOnChanges.call(this, changes);
            this.setupChanges('allowedFileExtensions', changes);
            this.setupChanges('currentPathKeys', changes);
            this.setupChanges('selectedItemKeys', changes);
        };
        DxFileManagerComponent.prototype.setupChanges = function (prop, changes) {
            if (!(prop in this._optionsToUpdate)) {
                this._idh.setup(prop, changes);
            }
        };
        DxFileManagerComponent.prototype.ngDoCheck = function () {
            this._idh.doCheck('allowedFileExtensions');
            this._idh.doCheck('currentPathKeys');
            this._idh.doCheck('selectedItemKeys');
            this._watcherHelper.checkWatchers();
            _super.prototype.ngDoCheck.call(this);
            _super.prototype.clearChangedOptions.call(this);
        };
        DxFileManagerComponent.prototype._setOption = function (name, value) {
            var isSetup = this._idh.setupSingle(name, value);
            var isChanged = this._idh.getChanges(name, value) !== null;
            if (isSetup || isChanged) {
                _super.prototype._setOption.call(this, name, value);
            }
        };
        return DxFileManagerComponent;
    }(i1.DxComponent));
    /** @nocollapse */ DxFileManagerComponent.ɵfac = function DxFileManagerComponent_Factory(t) { return new (t || DxFileManagerComponent)(i0__namespace.ɵɵdirectiveInject(i0__namespace.ElementRef), i0__namespace.ɵɵdirectiveInject(i0__namespace.NgZone), i0__namespace.ɵɵdirectiveInject(i1__namespace.DxTemplateHost), i0__namespace.ɵɵdirectiveInject(i1__namespace.WatcherHelper), i0__namespace.ɵɵdirectiveInject(i1__namespace.IterableDifferHelper), i0__namespace.ɵɵdirectiveInject(i1__namespace.NestedOptionHost), i0__namespace.ɵɵdirectiveInject(i2__namespace.TransferState), i0__namespace.ɵɵdirectiveInject(i0.PLATFORM_ID)); };
    /** @nocollapse */ DxFileManagerComponent.ɵcmp = /** @pureOrBreakMyCode */ i0__namespace.ɵɵdefineComponent({ type: DxFileManagerComponent, selectors: [["dx-file-manager"]], inputs: { accessKey: "accessKey", activeStateEnabled: "activeStateEnabled", allowedFileExtensions: "allowedFileExtensions", contextMenu: "contextMenu", currentPath: "currentPath", currentPathKeys: "currentPathKeys", customizeDetailColumns: "customizeDetailColumns", customizeThumbnail: "customizeThumbnail", disabled: "disabled", elementAttr: "elementAttr", fileSystemProvider: "fileSystemProvider", focusedItemKey: "focusedItemKey", focusStateEnabled: "focusStateEnabled", height: "height", hint: "hint", hoverStateEnabled: "hoverStateEnabled", itemView: "itemView", notifications: "notifications", permissions: "permissions", rootFolderName: "rootFolderName", rtlEnabled: "rtlEnabled", selectedItemKeys: "selectedItemKeys", selectionMode: "selectionMode", tabIndex: "tabIndex", toolbar: "toolbar", upload: "upload", visible: "visible", width: "width" }, outputs: { onContentReady: "onContentReady", onContextMenuItemClick: "onContextMenuItemClick", onContextMenuShowing: "onContextMenuShowing", onCurrentDirectoryChanged: "onCurrentDirectoryChanged", onDirectoryCreated: "onDirectoryCreated", onDirectoryCreating: "onDirectoryCreating", onDisposing: "onDisposing", onErrorOccurred: "onErrorOccurred", onFileUploaded: "onFileUploaded", onFileUploading: "onFileUploading", onFocusedItemChanged: "onFocusedItemChanged", onInitialized: "onInitialized", onItemCopied: "onItemCopied", onItemCopying: "onItemCopying", onItemDeleted: "onItemDeleted", onItemDeleting: "onItemDeleting", onItemDownloading: "onItemDownloading", onItemMoved: "onItemMoved", onItemMoving: "onItemMoving", onItemRenamed: "onItemRenamed", onItemRenaming: "onItemRenaming", onOptionChanged: "onOptionChanged", onSelectedFileOpened: "onSelectedFileOpened", onSelectionChanged: "onSelectionChanged", onToolbarItemClick: "onToolbarItemClick", accessKeyChange: "accessKeyChange", activeStateEnabledChange: "activeStateEnabledChange", allowedFileExtensionsChange: "allowedFileExtensionsChange", contextMenuChange: "contextMenuChange", currentPathChange: "currentPathChange", currentPathKeysChange: "currentPathKeysChange", customizeDetailColumnsChange: "customizeDetailColumnsChange", customizeThumbnailChange: "customizeThumbnailChange", disabledChange: "disabledChange", elementAttrChange: "elementAttrChange", fileSystemProviderChange: "fileSystemProviderChange", focusedItemKeyChange: "focusedItemKeyChange", focusStateEnabledChange: "focusStateEnabledChange", heightChange: "heightChange", hintChange: "hintChange", hoverStateEnabledChange: "hoverStateEnabledChange", itemViewChange: "itemViewChange", notificationsChange: "notificationsChange", permissionsChange: "permissionsChange", rootFolderNameChange: "rootFolderNameChange", rtlEnabledChange: "rtlEnabledChange", selectedItemKeysChange: "selectedItemKeysChange", selectionModeChange: "selectionModeChange", tabIndexChange: "tabIndexChange", toolbarChange: "toolbarChange", uploadChange: "uploadChange", visibleChange: "visibleChange", widthChange: "widthChange" }, features: [i0__namespace.ɵɵProvidersFeature([
                i1.DxTemplateHost,
                i1.WatcherHelper,
                i1.NestedOptionHost,
                i1.IterableDifferHelper
            ]), i0__namespace.ɵɵInheritDefinitionFeature, i0__namespace.ɵɵNgOnChangesFeature], decls: 0, vars: 0, template: function DxFileManagerComponent_Template(rf, ctx) { }, encapsulation: 2 });
    (function () {
        (typeof ngDevMode === "undefined" || ngDevMode) && i0__namespace.ɵsetClassMetadata(DxFileManagerComponent, [{
                type: i0.Component,
                args: [{
                        selector: 'dx-file-manager',
                        template: '',
                        providers: [
                            i1.DxTemplateHost,
                            i1.WatcherHelper,
                            i1.NestedOptionHost,
                            i1.IterableDifferHelper
                        ]
                    }]
            }], function () {
            return [{ type: i0__namespace.ElementRef }, { type: i0__namespace.NgZone }, { type: i1__namespace.DxTemplateHost }, { type: i1__namespace.WatcherHelper }, { type: i1__namespace.IterableDifferHelper }, { type: i1__namespace.NestedOptionHost }, { type: i2__namespace.TransferState }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i0.PLATFORM_ID]
                        }] }];
        }, { accessKey: [{
                    type: i0.Input
                }], activeStateEnabled: [{
                    type: i0.Input
                }], allowedFileExtensions: [{
                    type: i0.Input
                }], contextMenu: [{
                    type: i0.Input
                }], currentPath: [{
                    type: i0.Input
                }], currentPathKeys: [{
                    type: i0.Input
                }], customizeDetailColumns: [{
                    type: i0.Input
                }], customizeThumbnail: [{
                    type: i0.Input
                }], disabled: [{
                    type: i0.Input
                }], elementAttr: [{
                    type: i0.Input
                }], fileSystemProvider: [{
                    type: i0.Input
                }], focusedItemKey: [{
                    type: i0.Input
                }], focusStateEnabled: [{
                    type: i0.Input
                }], height: [{
                    type: i0.Input
                }], hint: [{
                    type: i0.Input
                }], hoverStateEnabled: [{
                    type: i0.Input
                }], itemView: [{
                    type: i0.Input
                }], notifications: [{
                    type: i0.Input
                }], permissions: [{
                    type: i0.Input
                }], rootFolderName: [{
                    type: i0.Input
                }], rtlEnabled: [{
                    type: i0.Input
                }], selectedItemKeys: [{
                    type: i0.Input
                }], selectionMode: [{
                    type: i0.Input
                }], tabIndex: [{
                    type: i0.Input
                }], toolbar: [{
                    type: i0.Input
                }], upload: [{
                    type: i0.Input
                }], visible: [{
                    type: i0.Input
                }], width: [{
                    type: i0.Input
                }], onContentReady: [{
                    type: i0.Output
                }], onContextMenuItemClick: [{
                    type: i0.Output
                }], onContextMenuShowing: [{
                    type: i0.Output
                }], onCurrentDirectoryChanged: [{
                    type: i0.Output
                }], onDirectoryCreated: [{
                    type: i0.Output
                }], onDirectoryCreating: [{
                    type: i0.Output
                }], onDisposing: [{
                    type: i0.Output
                }], onErrorOccurred: [{
                    type: i0.Output
                }], onFileUploaded: [{
                    type: i0.Output
                }], onFileUploading: [{
                    type: i0.Output
                }], onFocusedItemChanged: [{
                    type: i0.Output
                }], onInitialized: [{
                    type: i0.Output
                }], onItemCopied: [{
                    type: i0.Output
                }], onItemCopying: [{
                    type: i0.Output
                }], onItemDeleted: [{
                    type: i0.Output
                }], onItemDeleting: [{
                    type: i0.Output
                }], onItemDownloading: [{
                    type: i0.Output
                }], onItemMoved: [{
                    type: i0.Output
                }], onItemMoving: [{
                    type: i0.Output
                }], onItemRenamed: [{
                    type: i0.Output
                }], onItemRenaming: [{
                    type: i0.Output
                }], onOptionChanged: [{
                    type: i0.Output
                }], onSelectedFileOpened: [{
                    type: i0.Output
                }], onSelectionChanged: [{
                    type: i0.Output
                }], onToolbarItemClick: [{
                    type: i0.Output
                }], accessKeyChange: [{
                    type: i0.Output
                }], activeStateEnabledChange: [{
                    type: i0.Output
                }], allowedFileExtensionsChange: [{
                    type: i0.Output
                }], contextMenuChange: [{
                    type: i0.Output
                }], currentPathChange: [{
                    type: i0.Output
                }], currentPathKeysChange: [{
                    type: i0.Output
                }], customizeDetailColumnsChange: [{
                    type: i0.Output
                }], customizeThumbnailChange: [{
                    type: i0.Output
                }], disabledChange: [{
                    type: i0.Output
                }], elementAttrChange: [{
                    type: i0.Output
                }], fileSystemProviderChange: [{
                    type: i0.Output
                }], focusedItemKeyChange: [{
                    type: i0.Output
                }], focusStateEnabledChange: [{
                    type: i0.Output
                }], heightChange: [{
                    type: i0.Output
                }], hintChange: [{
                    type: i0.Output
                }], hoverStateEnabledChange: [{
                    type: i0.Output
                }], itemViewChange: [{
                    type: i0.Output
                }], notificationsChange: [{
                    type: i0.Output
                }], permissionsChange: [{
                    type: i0.Output
                }], rootFolderNameChange: [{
                    type: i0.Output
                }], rtlEnabledChange: [{
                    type: i0.Output
                }], selectedItemKeysChange: [{
                    type: i0.Output
                }], selectionModeChange: [{
                    type: i0.Output
                }], tabIndexChange: [{
                    type: i0.Output
                }], toolbarChange: [{
                    type: i0.Output
                }], uploadChange: [{
                    type: i0.Output
                }], visibleChange: [{
                    type: i0.Output
                }], widthChange: [{
                    type: i0.Output
                }] });
    })();
    var DxFileManagerModule = /** @class */ (function () {
        function DxFileManagerModule() {
        }
        return DxFileManagerModule;
    }());
    /** @nocollapse */ DxFileManagerModule.ɵfac = function DxFileManagerModule_Factory(t) { return new (t || DxFileManagerModule)(); };
    /** @nocollapse */ DxFileManagerModule.ɵmod = /** @pureOrBreakMyCode */ i0__namespace.ɵɵdefineNgModule({ type: DxFileManagerModule });
    /** @nocollapse */ DxFileManagerModule.ɵinj = /** @pureOrBreakMyCode */ i0__namespace.ɵɵdefineInjector({ imports: [[
                nested.DxoContextMenuModule,
                nested.DxiItemModule,
                nested.DxoItemViewModule,
                nested.DxoDetailsModule,
                nested.DxiColumnModule,
                nested.DxoNotificationsModule,
                nested.DxoPermissionsModule,
                nested.DxoToolbarModule,
                nested.DxiFileSelectionItemModule,
                nested.DxoUploadModule,
                i1.DxIntegrationModule,
                i1.DxTemplateModule,
                i2.BrowserTransferStateModule
            ], nested.DxoContextMenuModule,
            nested.DxiItemModule,
            nested.DxoItemViewModule,
            nested.DxoDetailsModule,
            nested.DxiColumnModule,
            nested.DxoNotificationsModule,
            nested.DxoPermissionsModule,
            nested.DxoToolbarModule,
            nested.DxiFileSelectionItemModule,
            nested.DxoUploadModule,
            i1.DxTemplateModule] });
    (function () {
        (typeof ngDevMode === "undefined" || ngDevMode) && i0__namespace.ɵsetClassMetadata(DxFileManagerModule, [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            nested.DxoContextMenuModule,
                            nested.DxiItemModule,
                            nested.DxoItemViewModule,
                            nested.DxoDetailsModule,
                            nested.DxiColumnModule,
                            nested.DxoNotificationsModule,
                            nested.DxoPermissionsModule,
                            nested.DxoToolbarModule,
                            nested.DxiFileSelectionItemModule,
                            nested.DxoUploadModule,
                            i1.DxIntegrationModule,
                            i1.DxTemplateModule,
                            i2.BrowserTransferStateModule
                        ],
                        declarations: [
                            DxFileManagerComponent
                        ],
                        exports: [
                            DxFileManagerComponent,
                            nested.DxoContextMenuModule,
                            nested.DxiItemModule,
                            nested.DxoItemViewModule,
                            nested.DxoDetailsModule,
                            nested.DxiColumnModule,
                            nested.DxoNotificationsModule,
                            nested.DxoPermissionsModule,
                            nested.DxoToolbarModule,
                            nested.DxiFileSelectionItemModule,
                            nested.DxoUploadModule,
                            i1.DxTemplateModule
                        ]
                    }]
            }], null, null);
    })();
    (function () {
        (typeof ngJitMode === "undefined" || ngJitMode) && i0__namespace.ɵɵsetNgModuleScope(DxFileManagerModule, { declarations: [DxFileManagerComponent], imports: [nested.DxoContextMenuModule,
                nested.DxiItemModule,
                nested.DxoItemViewModule,
                nested.DxoDetailsModule,
                nested.DxiColumnModule,
                nested.DxoNotificationsModule,
                nested.DxoPermissionsModule,
                nested.DxoToolbarModule,
                nested.DxiFileSelectionItemModule,
                nested.DxoUploadModule,
                i1.DxIntegrationModule,
                i1.DxTemplateModule,
                i2.BrowserTransferStateModule], exports: [DxFileManagerComponent, nested.DxoContextMenuModule,
                nested.DxiItemModule,
                nested.DxoItemViewModule,
                nested.DxoDetailsModule,
                nested.DxiColumnModule,
                nested.DxoNotificationsModule,
                nested.DxoPermissionsModule,
                nested.DxoToolbarModule,
                nested.DxiFileSelectionItemModule,
                nested.DxoUploadModule,
                i1.DxTemplateModule] });
    })();

    /**
     * Generated bundle index. Do not edit.
     */

    exports.DxFileManagerComponent = DxFileManagerComponent;
    exports.DxFileManagerModule = DxFileManagerModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
