// devextreme-angular/bundles/devextreme-angular-ui-popover.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/platform-browser'), require('@angular/core'), require('devextreme/ui/popover'), require('devextreme-angular/core'), require('devextreme-angular/ui/nested')) :
    typeof define === 'function' && define.amd ? define('devextreme-angular/ui/popover', ['exports', '@angular/platform-browser', '@angular/core', 'devextreme/ui/popover', 'devextreme-angular/core', 'devextreme-angular/ui/nested'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["devextreme-angular"] = global["devextreme-angular"] || {}, global["devextreme-angular"].ui = global["devextreme-angular"].ui || {}, global["devextreme-angular"].ui.popover = {}), global.ng.platformBrowser, global.ng.core, global.DxPopover, global["devextreme-angular"].core, global["devextreme-angular"].ui.nested));
})(this, (function (exports, i2, i0, DxPopover, i1, nested) { 'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var DxPopover__default = /*#__PURE__*/_interopDefaultLegacy(DxPopover);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    function __createBinding(o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    }
    function __exportStar(m, exports) {
        for (var p in m)
            if (p !== "default" && !exports.hasOwnProperty(p))
                exports[p] = m[p];
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (Object.hasOwnProperty.call(mod, k))
                    result[k] = mod[k];
        result.default = mod;
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var _c0 = ["*"];
    /**
     * The Popover is a UI component that shows notifications within a box with an arrow pointing to a specified UI element.

     */
    var DxPopoverComponent = /** @class */ (function (_super) {
        __extends(DxPopoverComponent, _super);
        function DxPopoverComponent(elementRef, ngZone, templateHost, _watcherHelper, _idh, optionHost, transferState, platformId) {
            var _this = _super.call(this, elementRef, ngZone, templateHost, _watcherHelper, transferState, platformId) || this;
            _this._watcherHelper = _watcherHelper;
            _this._idh = _idh;
            _this._createEventEmitters([
                { subscribe: 'contentReady', emit: 'onContentReady' },
                { subscribe: 'disposing', emit: 'onDisposing' },
                { subscribe: 'hidden', emit: 'onHidden' },
                { subscribe: 'hiding', emit: 'onHiding' },
                { subscribe: 'initialized', emit: 'onInitialized' },
                { subscribe: 'optionChanged', emit: 'onOptionChanged' },
                { subscribe: 'showing', emit: 'onShowing' },
                { subscribe: 'shown', emit: 'onShown' },
                { subscribe: 'titleRendered', emit: 'onTitleRendered' },
                { emit: 'animationChange' },
                { emit: 'closeOnOutsideClickChange' },
                { emit: 'containerChange' },
                { emit: 'contentTemplateChange' },
                { emit: 'copyRootClassesToWrapperChange' },
                { emit: 'deferRenderingChange' },
                { emit: 'disabledChange' },
                { emit: 'elementAttrChange' },
                { emit: 'heightChange' },
                { emit: 'hideEventChange' },
                { emit: 'hideOnOutsideClickChange' },
                { emit: 'hideOnParentScrollChange' },
                { emit: 'hintChange' },
                { emit: 'hoverStateEnabledChange' },
                { emit: 'maxHeightChange' },
                { emit: 'maxWidthChange' },
                { emit: 'minHeightChange' },
                { emit: 'minWidthChange' },
                { emit: 'positionChange' },
                { emit: 'rtlEnabledChange' },
                { emit: 'shadingChange' },
                { emit: 'shadingColorChange' },
                { emit: 'showCloseButtonChange' },
                { emit: 'showEventChange' },
                { emit: 'showTitleChange' },
                { emit: 'targetChange' },
                { emit: 'titleChange' },
                { emit: 'titleTemplateChange' },
                { emit: 'toolbarItemsChange' },
                { emit: 'visibleChange' },
                { emit: 'widthChange' },
                { emit: 'wrapperAttrChange' }
            ]);
            _this._idh.setHost(_this);
            optionHost.setHost(_this);
            return _this;
        }
        Object.defineProperty(DxPopoverComponent.prototype, "animation", {
            /**
             * Configures UI component visibility animations. This object contains two fields: show and hide.
            
             */
            get: function () {
                return this._getOption('animation');
            },
            set: function (value) {
                this._setOption('animation', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "closeOnOutsideClick", {
            /**
             * Specifies whether to close the UI component if a user clicks outside the popover window or outside the target element.
            
             * @deprecated Use the hideOnOutsideClick option instead.
            
             */
            get: function () {
                return this._getOption('closeOnOutsideClick');
            },
            set: function (value) {
                this._setOption('closeOnOutsideClick', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "container", {
            /**
             * Specifies the container in which to render the UI component.
            
             */
            get: function () {
                return this._getOption('container');
            },
            set: function (value) {
                this._setOption('container', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "contentTemplate", {
            /**
             * Specifies a custom template for the UI component content.
            
             */
            get: function () {
                return this._getOption('contentTemplate');
            },
            set: function (value) {
                this._setOption('contentTemplate', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "copyRootClassesToWrapper", {
            /**
             * Copies your custom CSS classes from the root element to the wrapper element.
            
             * @deprecated Use the wrapperAttr property instead.
            
             */
            get: function () {
                return this._getOption('copyRootClassesToWrapper');
            },
            set: function (value) {
                this._setOption('copyRootClassesToWrapper', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "deferRendering", {
            /**
             * Specifies whether to render the UI component&apos;s content when it is displayed. If false, the content is rendered immediately.
            
             */
            get: function () {
                return this._getOption('deferRendering');
            },
            set: function (value) {
                this._setOption('deferRendering', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "disabled", {
            /**
             * Specifies whether the UI component responds to user interaction.
            
             */
            get: function () {
                return this._getOption('disabled');
            },
            set: function (value) {
                this._setOption('disabled', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "elementAttr", {
            /**
             * Specifies the global attributes to be attached to the UI component&apos;s container element.
            
             * @deprecated Use the wrapperAttr option instead.
            
             */
            get: function () {
                return this._getOption('elementAttr');
            },
            set: function (value) {
                this._setOption('elementAttr', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "height", {
            /**
             * Specifies the UI component&apos;s height.
            
             */
            get: function () {
                return this._getOption('height');
            },
            set: function (value) {
                this._setOption('height', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "hideEvent", {
            /**
             * Specifies properties of popover hiding. Ignored if the shading property is set to true.
            
             */
            get: function () {
                return this._getOption('hideEvent');
            },
            set: function (value) {
                this._setOption('hideEvent', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "hideOnOutsideClick", {
            /**
             * Specifies whether to hide the UI component if a user clicks outside the popover window or outside the target element.
            
             */
            get: function () {
                return this._getOption('hideOnOutsideClick');
            },
            set: function (value) {
                this._setOption('hideOnOutsideClick', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "hideOnParentScroll", {
            /**
             * Specifies whether to hide the widget when users scroll one of its parent elements.
            
             */
            get: function () {
                return this._getOption('hideOnParentScroll');
            },
            set: function (value) {
                this._setOption('hideOnParentScroll', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "hint", {
            /**
             * Specifies text for a hint that appears when a user pauses on the UI component.
            
             */
            get: function () {
                return this._getOption('hint');
            },
            set: function (value) {
                this._setOption('hint', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "hoverStateEnabled", {
            /**
             * Specifies whether the UI component changes its state when a user pauses on it.
            
             */
            get: function () {
                return this._getOption('hoverStateEnabled');
            },
            set: function (value) {
                this._setOption('hoverStateEnabled', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "maxHeight", {
            /**
             * Specifies the maximum height the UI component can reach while resizing.
            
             */
            get: function () {
                return this._getOption('maxHeight');
            },
            set: function (value) {
                this._setOption('maxHeight', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "maxWidth", {
            /**
             * Specifies the maximum width the UI component can reach while resizing.
            
             */
            get: function () {
                return this._getOption('maxWidth');
            },
            set: function (value) {
                this._setOption('maxWidth', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "minHeight", {
            /**
             * Specifies the minimum height the UI component can reach while resizing.
            
             */
            get: function () {
                return this._getOption('minHeight');
            },
            set: function (value) {
                this._setOption('minHeight', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "minWidth", {
            /**
             * Specifies the minimum width the UI component can reach while resizing.
            
             */
            get: function () {
                return this._getOption('minWidth');
            },
            set: function (value) {
                this._setOption('minWidth', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "position", {
            /**
             * An object defining UI component positioning properties.
            
             */
            get: function () {
                return this._getOption('position');
            },
            set: function (value) {
                this._setOption('position', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "rtlEnabled", {
            /**
             * Switches the UI component to a right-to-left representation.
            
             */
            get: function () {
                return this._getOption('rtlEnabled');
            },
            set: function (value) {
                this._setOption('rtlEnabled', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "shading", {
            /**
             * Specifies whether to shade the background when the UI component is active.
            
             */
            get: function () {
                return this._getOption('shading');
            },
            set: function (value) {
                this._setOption('shading', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "shadingColor", {
            /**
             * Specifies the shading color. Applies only if shading is enabled.
            
             */
            get: function () {
                return this._getOption('shadingColor');
            },
            set: function (value) {
                this._setOption('shadingColor', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "showCloseButton", {
            /**
             * Specifies whether or not the UI component displays the Close button.
            
             */
            get: function () {
                return this._getOption('showCloseButton');
            },
            set: function (value) {
                this._setOption('showCloseButton', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "showEvent", {
            /**
             * Specifies properties for displaying the UI component.
            
             */
            get: function () {
                return this._getOption('showEvent');
            },
            set: function (value) {
                this._setOption('showEvent', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "showTitle", {
            /**
             * A Boolean value specifying whether or not to display the title in the overlay window.
            
             */
            get: function () {
                return this._getOption('showTitle');
            },
            set: function (value) {
                this._setOption('showTitle', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "target", {
            /**
             * Specifies the element against which to position the widget.
            
             */
            get: function () {
                return this._getOption('target');
            },
            set: function (value) {
                this._setOption('target', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "title", {
            /**
             * The title in the overlay window.
            
             */
            get: function () {
                return this._getOption('title');
            },
            set: function (value) {
                this._setOption('title', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "titleTemplate", {
            /**
             * Specifies a custom template for the UI component title. Does not apply if the title is defined.
            
             */
            get: function () {
                return this._getOption('titleTemplate');
            },
            set: function (value) {
                this._setOption('titleTemplate', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "toolbarItems", {
            /**
             * Configures toolbar items.
            
             */
            get: function () {
                return this._getOption('toolbarItems');
            },
            set: function (value) {
                this._setOption('toolbarItems', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "visible", {
            /**
             * A Boolean value specifying whether or not the UI component is visible.
            
             */
            get: function () {
                return this._getOption('visible');
            },
            set: function (value) {
                this._setOption('visible', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "width", {
            /**
             * Specifies the UI component&apos;s width.
            
             */
            get: function () {
                return this._getOption('width');
            },
            set: function (value) {
                this._setOption('width', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "wrapperAttr", {
            /**
             * Specifies the global attributes for the UI component&apos;s wrapper element.
            
             */
            get: function () {
                return this._getOption('wrapperAttr');
            },
            set: function (value) {
                this._setOption('wrapperAttr', value);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DxPopoverComponent.prototype, "toolbarItemsChildren", {
            get: function () {
                return this._getOption('toolbarItems');
            },
            set: function (value) {
                this.setChildren('toolbarItems', value);
            },
            enumerable: false,
            configurable: true
        });
        DxPopoverComponent.prototype._createInstance = function (element, options) {
            return new DxPopover__default["default"](element, options);
        };
        DxPopoverComponent.prototype.ngOnDestroy = function () {
            this._destroyWidget();
        };
        DxPopoverComponent.prototype.ngOnChanges = function (changes) {
            _super.prototype.ngOnChanges.call(this, changes);
            this.setupChanges('toolbarItems', changes);
        };
        DxPopoverComponent.prototype.setupChanges = function (prop, changes) {
            if (!(prop in this._optionsToUpdate)) {
                this._idh.setup(prop, changes);
            }
        };
        DxPopoverComponent.prototype.ngDoCheck = function () {
            this._idh.doCheck('toolbarItems');
            this._watcherHelper.checkWatchers();
            _super.prototype.ngDoCheck.call(this);
            _super.prototype.clearChangedOptions.call(this);
        };
        DxPopoverComponent.prototype._setOption = function (name, value) {
            var isSetup = this._idh.setupSingle(name, value);
            var isChanged = this._idh.getChanges(name, value) !== null;
            if (isSetup || isChanged) {
                _super.prototype._setOption.call(this, name, value);
            }
        };
        return DxPopoverComponent;
    }(i1.DxComponent));
    /** @nocollapse */ DxPopoverComponent.ɵfac = function DxPopoverComponent_Factory(t) { return new (t || DxPopoverComponent)(i0__namespace.ɵɵdirectiveInject(i0__namespace.ElementRef), i0__namespace.ɵɵdirectiveInject(i0__namespace.NgZone), i0__namespace.ɵɵdirectiveInject(i1__namespace.DxTemplateHost), i0__namespace.ɵɵdirectiveInject(i1__namespace.WatcherHelper), i0__namespace.ɵɵdirectiveInject(i1__namespace.IterableDifferHelper), i0__namespace.ɵɵdirectiveInject(i1__namespace.NestedOptionHost), i0__namespace.ɵɵdirectiveInject(i2__namespace.TransferState), i0__namespace.ɵɵdirectiveInject(i0.PLATFORM_ID)); };
    /** @nocollapse */ DxPopoverComponent.ɵcmp = /** @pureOrBreakMyCode */ i0__namespace.ɵɵdefineComponent({ type: DxPopoverComponent, selectors: [["dx-popover"]], contentQueries: function DxPopoverComponent_ContentQueries(rf, ctx, dirIndex) {
            if (rf & 1) {
                i0__namespace.ɵɵcontentQuery(dirIndex, nested.DxiToolbarItemComponent, 4);
            }
            if (rf & 2) {
                var _t = void 0;
                i0__namespace.ɵɵqueryRefresh(_t = i0__namespace.ɵɵloadQuery()) && (ctx.toolbarItemsChildren = _t);
            }
        }, inputs: { animation: "animation", closeOnOutsideClick: "closeOnOutsideClick", container: "container", contentTemplate: "contentTemplate", copyRootClassesToWrapper: "copyRootClassesToWrapper", deferRendering: "deferRendering", disabled: "disabled", elementAttr: "elementAttr", height: "height", hideEvent: "hideEvent", hideOnOutsideClick: "hideOnOutsideClick", hideOnParentScroll: "hideOnParentScroll", hint: "hint", hoverStateEnabled: "hoverStateEnabled", maxHeight: "maxHeight", maxWidth: "maxWidth", minHeight: "minHeight", minWidth: "minWidth", position: "position", rtlEnabled: "rtlEnabled", shading: "shading", shadingColor: "shadingColor", showCloseButton: "showCloseButton", showEvent: "showEvent", showTitle: "showTitle", target: "target", title: "title", titleTemplate: "titleTemplate", toolbarItems: "toolbarItems", visible: "visible", width: "width", wrapperAttr: "wrapperAttr" }, outputs: { onContentReady: "onContentReady", onDisposing: "onDisposing", onHidden: "onHidden", onHiding: "onHiding", onInitialized: "onInitialized", onOptionChanged: "onOptionChanged", onShowing: "onShowing", onShown: "onShown", onTitleRendered: "onTitleRendered", animationChange: "animationChange", closeOnOutsideClickChange: "closeOnOutsideClickChange", containerChange: "containerChange", contentTemplateChange: "contentTemplateChange", copyRootClassesToWrapperChange: "copyRootClassesToWrapperChange", deferRenderingChange: "deferRenderingChange", disabledChange: "disabledChange", elementAttrChange: "elementAttrChange", heightChange: "heightChange", hideEventChange: "hideEventChange", hideOnOutsideClickChange: "hideOnOutsideClickChange", hideOnParentScrollChange: "hideOnParentScrollChange", hintChange: "hintChange", hoverStateEnabledChange: "hoverStateEnabledChange", maxHeightChange: "maxHeightChange", maxWidthChange: "maxWidthChange", minHeightChange: "minHeightChange", minWidthChange: "minWidthChange", positionChange: "positionChange", rtlEnabledChange: "rtlEnabledChange", shadingChange: "shadingChange", shadingColorChange: "shadingColorChange", showCloseButtonChange: "showCloseButtonChange", showEventChange: "showEventChange", showTitleChange: "showTitleChange", targetChange: "targetChange", titleChange: "titleChange", titleTemplateChange: "titleTemplateChange", toolbarItemsChange: "toolbarItemsChange", visibleChange: "visibleChange", widthChange: "widthChange", wrapperAttrChange: "wrapperAttrChange" }, features: [i0__namespace.ɵɵProvidersFeature([
                i1.DxTemplateHost,
                i1.WatcherHelper,
                i1.NestedOptionHost,
                i1.IterableDifferHelper
            ]), i0__namespace.ɵɵInheritDefinitionFeature, i0__namespace.ɵɵNgOnChangesFeature], ngContentSelectors: _c0, decls: 1, vars: 0, template: function DxPopoverComponent_Template(rf, ctx) {
            if (rf & 1) {
                i0__namespace.ɵɵprojectionDef();
                i0__namespace.ɵɵprojection(0);
            }
        }, encapsulation: 2 });
    (function () {
        (typeof ngDevMode === "undefined" || ngDevMode) && i0__namespace.ɵsetClassMetadata(DxPopoverComponent, [{
                type: i0.Component,
                args: [{
                        selector: 'dx-popover',
                        template: '<ng-content></ng-content>',
                        providers: [
                            i1.DxTemplateHost,
                            i1.WatcherHelper,
                            i1.NestedOptionHost,
                            i1.IterableDifferHelper
                        ]
                    }]
            }], function () {
            return [{ type: i0__namespace.ElementRef }, { type: i0__namespace.NgZone }, { type: i1__namespace.DxTemplateHost }, { type: i1__namespace.WatcherHelper }, { type: i1__namespace.IterableDifferHelper }, { type: i1__namespace.NestedOptionHost }, { type: i2__namespace.TransferState }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i0.PLATFORM_ID]
                        }] }];
        }, { animation: [{
                    type: i0.Input
                }], closeOnOutsideClick: [{
                    type: i0.Input
                }], container: [{
                    type: i0.Input
                }], contentTemplate: [{
                    type: i0.Input
                }], copyRootClassesToWrapper: [{
                    type: i0.Input
                }], deferRendering: [{
                    type: i0.Input
                }], disabled: [{
                    type: i0.Input
                }], elementAttr: [{
                    type: i0.Input
                }], height: [{
                    type: i0.Input
                }], hideEvent: [{
                    type: i0.Input
                }], hideOnOutsideClick: [{
                    type: i0.Input
                }], hideOnParentScroll: [{
                    type: i0.Input
                }], hint: [{
                    type: i0.Input
                }], hoverStateEnabled: [{
                    type: i0.Input
                }], maxHeight: [{
                    type: i0.Input
                }], maxWidth: [{
                    type: i0.Input
                }], minHeight: [{
                    type: i0.Input
                }], minWidth: [{
                    type: i0.Input
                }], position: [{
                    type: i0.Input
                }], rtlEnabled: [{
                    type: i0.Input
                }], shading: [{
                    type: i0.Input
                }], shadingColor: [{
                    type: i0.Input
                }], showCloseButton: [{
                    type: i0.Input
                }], showEvent: [{
                    type: i0.Input
                }], showTitle: [{
                    type: i0.Input
                }], target: [{
                    type: i0.Input
                }], title: [{
                    type: i0.Input
                }], titleTemplate: [{
                    type: i0.Input
                }], toolbarItems: [{
                    type: i0.Input
                }], visible: [{
                    type: i0.Input
                }], width: [{
                    type: i0.Input
                }], wrapperAttr: [{
                    type: i0.Input
                }], onContentReady: [{
                    type: i0.Output
                }], onDisposing: [{
                    type: i0.Output
                }], onHidden: [{
                    type: i0.Output
                }], onHiding: [{
                    type: i0.Output
                }], onInitialized: [{
                    type: i0.Output
                }], onOptionChanged: [{
                    type: i0.Output
                }], onShowing: [{
                    type: i0.Output
                }], onShown: [{
                    type: i0.Output
                }], onTitleRendered: [{
                    type: i0.Output
                }], animationChange: [{
                    type: i0.Output
                }], closeOnOutsideClickChange: [{
                    type: i0.Output
                }], containerChange: [{
                    type: i0.Output
                }], contentTemplateChange: [{
                    type: i0.Output
                }], copyRootClassesToWrapperChange: [{
                    type: i0.Output
                }], deferRenderingChange: [{
                    type: i0.Output
                }], disabledChange: [{
                    type: i0.Output
                }], elementAttrChange: [{
                    type: i0.Output
                }], heightChange: [{
                    type: i0.Output
                }], hideEventChange: [{
                    type: i0.Output
                }], hideOnOutsideClickChange: [{
                    type: i0.Output
                }], hideOnParentScrollChange: [{
                    type: i0.Output
                }], hintChange: [{
                    type: i0.Output
                }], hoverStateEnabledChange: [{
                    type: i0.Output
                }], maxHeightChange: [{
                    type: i0.Output
                }], maxWidthChange: [{
                    type: i0.Output
                }], minHeightChange: [{
                    type: i0.Output
                }], minWidthChange: [{
                    type: i0.Output
                }], positionChange: [{
                    type: i0.Output
                }], rtlEnabledChange: [{
                    type: i0.Output
                }], shadingChange: [{
                    type: i0.Output
                }], shadingColorChange: [{
                    type: i0.Output
                }], showCloseButtonChange: [{
                    type: i0.Output
                }], showEventChange: [{
                    type: i0.Output
                }], showTitleChange: [{
                    type: i0.Output
                }], targetChange: [{
                    type: i0.Output
                }], titleChange: [{
                    type: i0.Output
                }], titleTemplateChange: [{
                    type: i0.Output
                }], toolbarItemsChange: [{
                    type: i0.Output
                }], visibleChange: [{
                    type: i0.Output
                }], widthChange: [{
                    type: i0.Output
                }], wrapperAttrChange: [{
                    type: i0.Output
                }], toolbarItemsChildren: [{
                    type: i0.ContentChildren,
                    args: [nested.DxiToolbarItemComponent]
                }] });
    })();
    var DxPopoverModule = /** @class */ (function () {
        function DxPopoverModule() {
        }
        return DxPopoverModule;
    }());
    /** @nocollapse */ DxPopoverModule.ɵfac = function DxPopoverModule_Factory(t) { return new (t || DxPopoverModule)(); };
    /** @nocollapse */ DxPopoverModule.ɵmod = /** @pureOrBreakMyCode */ i0__namespace.ɵɵdefineNgModule({ type: DxPopoverModule });
    /** @nocollapse */ DxPopoverModule.ɵinj = /** @pureOrBreakMyCode */ i0__namespace.ɵɵdefineInjector({ imports: [[
                nested.DxoAnimationModule,
                nested.DxoHideModule,
                nested.DxoFromModule,
                nested.DxoPositionModule,
                nested.DxoAtModule,
                nested.DxoBoundaryOffsetModule,
                nested.DxoCollisionModule,
                nested.DxoMyModule,
                nested.DxoOffsetModule,
                nested.DxoToModule,
                nested.DxoShowModule,
                nested.DxoHideEventModule,
                nested.DxoShowEventModule,
                nested.DxiToolbarItemModule,
                i1.DxIntegrationModule,
                i1.DxTemplateModule,
                i2.BrowserTransferStateModule
            ], nested.DxoAnimationModule,
            nested.DxoHideModule,
            nested.DxoFromModule,
            nested.DxoPositionModule,
            nested.DxoAtModule,
            nested.DxoBoundaryOffsetModule,
            nested.DxoCollisionModule,
            nested.DxoMyModule,
            nested.DxoOffsetModule,
            nested.DxoToModule,
            nested.DxoShowModule,
            nested.DxoHideEventModule,
            nested.DxoShowEventModule,
            nested.DxiToolbarItemModule,
            i1.DxTemplateModule] });
    (function () {
        (typeof ngDevMode === "undefined" || ngDevMode) && i0__namespace.ɵsetClassMetadata(DxPopoverModule, [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            nested.DxoAnimationModule,
                            nested.DxoHideModule,
                            nested.DxoFromModule,
                            nested.DxoPositionModule,
                            nested.DxoAtModule,
                            nested.DxoBoundaryOffsetModule,
                            nested.DxoCollisionModule,
                            nested.DxoMyModule,
                            nested.DxoOffsetModule,
                            nested.DxoToModule,
                            nested.DxoShowModule,
                            nested.DxoHideEventModule,
                            nested.DxoShowEventModule,
                            nested.DxiToolbarItemModule,
                            i1.DxIntegrationModule,
                            i1.DxTemplateModule,
                            i2.BrowserTransferStateModule
                        ],
                        declarations: [
                            DxPopoverComponent
                        ],
                        exports: [
                            DxPopoverComponent,
                            nested.DxoAnimationModule,
                            nested.DxoHideModule,
                            nested.DxoFromModule,
                            nested.DxoPositionModule,
                            nested.DxoAtModule,
                            nested.DxoBoundaryOffsetModule,
                            nested.DxoCollisionModule,
                            nested.DxoMyModule,
                            nested.DxoOffsetModule,
                            nested.DxoToModule,
                            nested.DxoShowModule,
                            nested.DxoHideEventModule,
                            nested.DxoShowEventModule,
                            nested.DxiToolbarItemModule,
                            i1.DxTemplateModule
                        ]
                    }]
            }], null, null);
    })();
    (function () {
        (typeof ngJitMode === "undefined" || ngJitMode) && i0__namespace.ɵɵsetNgModuleScope(DxPopoverModule, { declarations: [DxPopoverComponent], imports: [nested.DxoAnimationModule,
                nested.DxoHideModule,
                nested.DxoFromModule,
                nested.DxoPositionModule,
                nested.DxoAtModule,
                nested.DxoBoundaryOffsetModule,
                nested.DxoCollisionModule,
                nested.DxoMyModule,
                nested.DxoOffsetModule,
                nested.DxoToModule,
                nested.DxoShowModule,
                nested.DxoHideEventModule,
                nested.DxoShowEventModule,
                nested.DxiToolbarItemModule,
                i1.DxIntegrationModule,
                i1.DxTemplateModule,
                i2.BrowserTransferStateModule], exports: [DxPopoverComponent, nested.DxoAnimationModule,
                nested.DxoHideModule,
                nested.DxoFromModule,
                nested.DxoPositionModule,
                nested.DxoAtModule,
                nested.DxoBoundaryOffsetModule,
                nested.DxoCollisionModule,
                nested.DxoMyModule,
                nested.DxoOffsetModule,
                nested.DxoToModule,
                nested.DxoShowModule,
                nested.DxoHideEventModule,
                nested.DxoShowEventModule,
                nested.DxiToolbarItemModule,
                i1.DxTemplateModule] });
    })();

    /**
     * Generated bundle index. Do not edit.
     */

    exports.DxPopoverComponent = DxPopoverComponent;
    exports.DxPopoverModule = DxPopoverModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
